script.js

```diff
@@ -0,0 +1,841 @@
+      const { value } = vKeyBoard.textArea;
+      const cursorLineBegin = value.substring(0, cursorPosition).lastIndexOf("\n") + 1;
+      const cursorLength = cursorPosition - cursorLineBegin;
+      let nextLineBegin = value.indexOf("\n", cursorPosition) + 1;
+      if (nextLineBegin < 1) nextLineBegin = value.length;
+      let nextLineLength = value.indexOf("\n", nextLineBegin) - nextLineBegin;
+      if (nextLineLength < 0) nextLineLength = value.length - nextLineBegin;
+      let newPosition = nextLineBegin
+        + (cursorLength > nextLineLength ? nextLineLength : cursorLength);
+      if (newPosition < 0) newPosition = value.length;
+      vKeyBoard.focusTextArea(newPosition);
+    },
+    code: "ArrowDown",
+  },
+  {
+    type: "functional",
+    initial: "→",
+    func() {
+      const cursorPositionStart = textArea.selectionStart;
+      vKeyBoard.focusTextArea(cursorPositionStart + 1);
+    },
+    code: "ArrowRight",
+  },
+  {
+    type: "functional",
+    initial: "Ctrl",
+    isClick: true,
+    func: (keyDom) => {
+      FUNCTIONAL_KEYS.ctrlRight = !FUNCTIONAL_KEYS.ctrlRight;
+      if (FUNCTIONAL_KEYS.ctrlRight) keyDom.classList.add("keyboard-line__button_pressed");
+      else keyDom.classList.remove("keyboard-line__button_pressed");
+    },
+    code: "ControlRight",
+  },
+]);
+
+function getType(key) {
+  if (key.type) return key.type;
+  return key.languages[window.localStorage.getItem("virtualKeyBoardLang")].type;
+}
+
+function getChar(key, type) {
+  let isShifted = getType(key) === "number"
+    ? FUNCTIONAL_KEYS.shift
+    : FUNCTIONAL_KEYS.shift || FUNCTIONAL_KEYS.caps;
+  if (type) {
+    isShifted = type === "shifted";
+  }
+  if (key.initial) return isShifted ? key.shifted : key.initial;
+  return isShifted
+    ? key.languages[window.localStorage.getItem("virtualKeyBoardLang")].shifted
+    : key.languages[window.localStorage.getItem("virtualKeyBoardLang")].initial;
+}
+
+arrayOfButtonKeys.forEach((line) => {
+  const keyboardLine = document.createElement("div");
+  keyboardLine.className = "keyboard-line";
+  keyboardPanel.appendChild(keyboardLine);
+  line.forEach((key) => {
+    const keyDom = document.createElement("div");
+    keyDom.className = "keyboard-line__button";
+    keyDom.setAttribute("key", key.code);
+
+    const mainText = document.createElement("div");
+    mainText.className = "keyboard-line__button-main-text";
+    if (getType(key) === "letter" || getType(key) === "functional") {
+      keyDom.classList.add("keyboard-line__button_single");
+    }
+    keyDom.appendChild(mainText);
+    mainText.innerText = getChar(key, "initial");
+
+    const shiftedText = document.createElement("div");
+    shiftedText.className = "keyboard-line__button-shifted-text";
+    keyDom.appendChild(shiftedText);
+    if (getType(key) === "number") {
+      shiftedText.innerText = getChar(key, "shifted");
+    }
+
+    keyDom.addEventListener("mousedown", () => {

```

Comment:
As I see you add event listener for every key separately - it is not very good approach. Event delegation is better. You just delegate event handling to the parent element of all keys and handle events in event listener by getting from event button, that was clicked.

Reply:
Click and mouse down events assigned only to the keyboard panel instead of each key button and was moved to class.

https://github.com/BlackTruth/virtual-keyboard/blob/6297a0817c03affb1948c9e73a6352687e6ff449/src/classes/VirtualKeyboard.js#L70-L73
https://github.com/BlackTruth/virtual-keyboard/blob/6297a0817c03affb1948c9e73a6352687e6ff449/src/classes/VirtualKeyboard.js#L335